Adding RiboDetector to MGS pipeline

lennijusten opened the pull request:
Initial implementation of a new stage called `ribocounts` that saves a text file to AWS for each sample with the number of non-rRNA reads in the sample. Tested on Bohl 2022 and Wu 2019. 

- I initially wanted to save the number of rRNA reads per sample to a text file, but by default RiboDetector outputs non-rRNA reads that "pass" the screening step. Saving the number of rRNA reads would require an additional counting step and I thought it might be easier to do this at the dashboard stage. 
- RiboDetector takes an argument `--len` that should be set to roughly the average length of the input reads. I've added an option to perform the avg read length calculation for each input file, but it requires an additional pass over the unzipped input files. Open to discuss how we can improve this step.

lennijusten commented:
~TODO: update dependencies section of readme with RiboDetector install instructions~

jeffkaufman commented on run.py:
```diff
@@ -197,6 +197,106 @@ def get_files(args, dirname, min_size=1, min_date=''):
 
+def ribocounts(args):
+    """Count the number of reads classified as non-rRNA by RiboDetector"""
+    calc_avg_read_length = False
```
I think we should always do this -- it shouldn't be that slow

jeffkaufman commented on run.py:
```diff
@@ -197,6 +197,106 @@ def get_files(args, dirname, min_size=1, min_date=''):
+            if "discarded" in potential_input: continue
+
+            # Number of output and input files must match 
+            tmp_fq_outputs = potential_input.replace(".gz", ".nonrrna.fq")
```
`tmp_fq_output`

jeffkaufman commented on run.py:
```diff
@@ -197,6 +197,106 @@ def get_files(args, dirname, min_size=1, min_date=''):
+
+            # Number of output and input files must match 
+            tmp_fq_outputs = potential_input.replace(".gz", ".nonrrna.fq")
+            tmp_fq_outputs_list = [tmp_fq_outputs]
```
`tmp_fq_outputs = `

jeffkaufman commented on run.py:
```diff
@@ -197,6 +197,106 @@ def get_files(args, dirname, min_size=1, min_date=''):
+                # Add .fq extensions to input files
+                inputs = [i.replace(".gz", ".fq.gz") for i in inputs]
+
+                # Optional: compute average read lengths. Note that if we implement this option we can count total reads here. For paired-end reads, average length is computed only from pair1 reads.
```
nit: wrap long lines

jeffkaufman commented on run.py:
```diff
@@ -197,6 +197,106 @@ def get_files(args, dirname, min_size=1, min_date=''):
+                # Optional: compute average read lengths. Note that if we implement this option we can count total reads here. For paired-end reads, average length is computed only from pair1 reads.
+                if calc_avg_read_length:
+                    print("Calculating average read length...")
+                    def calculate_average_read_length(file_path):
```
Don't parse FASTQ manually in python; the format is a bit more subtle than this (ex: it allows wrapping long lines).  How about:

```
total_len = 0
total_reads = 0
with gzip.open(file_path, 'rt') as inf:
  for (title, sequence, quality) in FastqGeneralIterator(inf):
     total_len += len(sequence)
     total_reads += 1
return total_reads, round(total_len / total_reads)
```

jeffkaufman commented on run.py:
```diff
@@ -197,6 +197,106 @@ def get_files(args, dirname, min_size=1, min_date=''):
+                    print("Calculating average read length...")
+                    def calculate_average_read_length(file_path):
+                        with open(file_path, 'r') as f:
+                            lines = f.readlines()
```
fyi this requires a ton of memory -- much better to process the file a line at a time than to read the whole thing into memory before doing anything with it

jeffkaufman commented on run.py:
```diff
@@ -197,6 +197,106 @@ def get_files(args, dirname, min_size=1, min_date=''):
+
+                    unzipped_input_fname = inputs[0].replace(".gz", "")
+                    with open(unzipped_input_fname, 'wb') as f_out:
+                        subprocess.check_call(["gzip", "-c", "-d", inputs[0]], stdout=f_out)
```
you don't need to unzip it -- `gzip.open(...)` works just like `open(...)` but with transparent unzipping

jeffkaufman commented on run.py:
```diff
@@ -197,6 +197,106 @@ def get_files(args, dirname, min_size=1, min_date=''):
+                subprocess.check_call(ribodetector_cmd)
+                
+                # Count the number of reads passing rRNA screening (i.e. classified as non-rRNA)
+                def count_reads_in_fastq(file_path):
```
use `FastqGeneralIterator` here too

jeffkaufman commented on run.py:
```diff
@@ -197,6 +197,106 @@ def get_files(args, dirname, min_size=1, min_date=''):
+        # Send count number to bucket
+        with tempdir("ribocounts", sample + " output") as workdir:
+            with open(sample_output_file, 'w') as file:
+                file.write(str(sample_nonrrna_count))
```
better to write a trailing newline -- the standard way of doing text files on unix includes them.  Otherwise if you later  do `cat foo.ribocounts.txt` it will mess up your prompt.

lennijusten commented:
@jeffkaufman I addressed the requested changes in [ec27488](https://github.com/naobservatory/mgs-pipeline/pull/16/commits/ec274889e2821bcc7228753523314940e0491227) and updated the README with a description of the output data. 

Note that I changed the output to be **the number of rRNA reads** instead of the number of non-rRNA reads. It was possible to do this with minimal additional computation since I tallied the total sample reads during the average read length step.

I will need to re-run Wu 2020 and Bohl 2022 with this new output configuration.

jeffkaufman commented on README.md:
```diff
@@ -101,6 +101,14 @@ Available files, and their formats:
+    * ex: SRR13167443.ribocounts.txt
+    * txt
+    * RiboDetector default output format
+        * FASTQ file with all the input reads classified as non-rRNA.
```
I think this isn't right?  Didn't you decide not to upload a FASTQ?

jeffkaufman commented on run.py:
```diff
@@ -197,6 +197,104 @@ def get_files(args, dirname, min_size=1, min_date=''):
+                    return total_reads
+                
+                # Count non-rRNA reads in sample. Paired-end reads are counted once
+                sample_nonrrna_count += count_reads_in_fastq(tmp_fq_outputs[0])
```
You could do:

```
total_nonrrna_reads, _ = calculate_average_read_length(tmp_fq_outputs[0])
sample_nonrrna_count += total_nonrrna_reads
```

Then you don't need `count_reads_in_fastq`, though it's slightly less clear.

lennijusten replied:
I agree the current formulation is a bit clunky but I'm going to leave as is. One reason: `calculate_average_read_length` opens a gzip file where the output of RiboDetector is just a fastq file.
